From 82a750f4b6c6738e83e47a167cc9bd89391f0eae Mon Sep 17 00:00:00 2001
From: Sanyog Kale <skale@habana.ai>
Date: Mon, 09 Aug 2021 08:54:28 +0300
Subject: [PATCH] [SW-50880] ppboot: greco: arc: move arc0 init to early init

Since ARC0 needs to be used to perform PCIe initialization, this
patch moves init of ARC0 from plat init to plat early init.

Change-Id: I82efbf1add402f4e5b70146020b0deb22fbef3e5
---

diff --git a/arch/greco/arc_elf32/common.c b/arch/greco/arc_elf32/common.c
index 86bab10..fec9fac 100644
--- a/arch/greco/arc_elf32/common.c
+++ b/arch/greco/arc_elf32/common.c
@@ -119,6 +119,20 @@
 	regmap_writel(0x0, mmPSOC_GLOBAL_CONF_BOOT_SEQ_EXT_LD);
 }
 
+static void init_secondary(void)
+{
+	halt_n_reset_arc(0);
+
+	/*
+	 * Program Reset vector address for ARC0 to ICCM 0x0
+	 * location where vector table first entry will be available and
+	 * then bring ARC0 out of reset. ARCs when out of reset
+	 * is by default in halt state.
+	 */
+	regmap_writel(ARC_RST_VEC_ADDR, mmPSOC_ARC0_CFG_RST_VEC_ADDR);
+	regmap_writel(ARC_RST_VALUE, mmPSOC_ARC0_CFG_ARC_RST);
+}
+
 void hl_plat_early_init(void)
 {
 	/* Functionality used to be done in BTL */
@@ -127,6 +141,8 @@
 	/* PSOC AXI drain */
 	axi_enable_drain();
 
+	init_secondary();
+
 	cpu_boot_dev_sts_clrset(CPU_BOOT_DEV_STS0_FW_IATU_CONF_EN, true);
 
 	/*
@@ -134,20 +150,6 @@
 	 * restart boot sequence.
 	 */
 	regmap_writel(0x770, mmPSOC_GLOBAL_CONF_BOOT_SEQ_RE_START);
-}
-
-static void init_arc(void)
-{
-	halt_n_reset_arc(0);
-
-	/*
-	 * Program Reset vector address for ARC0 and ARC1 to ICCM 0x0
-	 * location where vector table first entry will be available and
-	 * then bring ARC0 and ARC1 out of reset. ARCs when out of reset
-	 * is by default in halt state.
-	 */
-	regmap_writel(ARC_RST_VEC_ADDR, mmPSOC_ARC0_CFG_RST_VEC_ADDR);
-	regmap_writel(ARC_RST_VALUE, mmPSOC_ARC0_CFG_ARC_RST);
 }
 
 #define SW_ALL_RST_L_MSK		0xC00FF8FF
@@ -163,8 +165,6 @@
 
 void hl_plat_init(void)
 {
-	init_arc();
-
 	sw_all_rst_init();
 }
 
